### PR TITLE
8225 HeaderManipulation Validation

### PR DESCRIPTION
**Pull request by sheidkamp**
# Description

We were passing through invalid values of `responseHeadersToAdd` to Envoy,  resulting in this type of error:
```
[2023-05-19 14:01:29.817][7][warning][config] [external/envoy/source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.route.v3.RouteConfiguration rejected: :-prefixed or host headers may not be modified
```

Steps to reproduce the bug
Create a VS with the following

    options:
      headerManipulation:
        responseHeadersToAdd:
          - header:
              key: ":scheme"
              value: "https"
            append: false

Additionally, we should not be allowing "Host" headers to be set (replace `:scheme` with `Host`)

# Context
Users want Gloo validation to catch this error instead passing invalid config to the gateway.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/main/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/pkg/plugins/headers/plugin_test.go [hunk 1]
```diff
@@ -111,6 +111,36 @@ var _ = Describe("Plugin", func() {
 		Expect(out.ResponseHeadersToAdd).To(Equal(expectedHeadersWithSecrets.ResponseHeadersToAdd))
 		Expect(out.ResponseHeadersToRemove).To(Equal(expectedHeadersWithSecrets.ResponseHeadersToRemove))
 	})
+
+	DescribeTable("Invalid headers", func(key string, value string, expectedErr error) {
```
Comment (elcasteel) [inline projects/gloo/pkg/plugins/headers/plugin_test.go:115]: 🎊

**Review by elcasteel — DISMISSED** [review]
LGTM. I think it makes sense to not attempt to do any type of "best effort" on the header manipulation if any of them are bad. 
Are the limitations documented?

File: projects/gloo/pkg/plugins/headers/plugin.go [hunk 2]
```diff
@@ -150,13 +157,23 @@ func convertResponseHeaderValueOption(
 ) ([]*envoy_config_core_v3.HeaderValueOption, error) {
 	var out []*envoy_config_core_v3.HeaderValueOption
 	for _, h := range in {
-		if h.GetHeader() == nil {
+		header := h.GetHeader()
+		if header == nil {
 			return nil, MissingHeaderValueError
 		}
+
+		if header.GetKey()[0] == byte(':') {
```
Comment (sam-heilbron) [inline projects/gloo/pkg/plugins/headers/plugin.go:165]: Any reason for not using the standard [HasPrefix](https://pkg.go.dev/strings#HasPrefix)?

File: projects/gloo/pkg/plugins/headers/plugin.go [hunk 3]
```diff
@@ -150,13 +157,23 @@ func convertResponseHeaderValueOption(
...
+			return nil, CantSetPseudoHeaderError(header.GetKey())
+		}
+
+		if strings.EqualFold(header.GetKey(), "Host") {
+			return nil, CantSetHostHeaderError
+		}
+
```
Comment (sam-heilbron) [inline projects/gloo/pkg/plugins/headers/plugin.go:172]: The logic in this space looks very similar to what we have in our api_conversion utility (https://github.com/solo-io/gloo/blob/main/pkg/utils/api_conversion/type.go#L54). Do these guards need to be added to those headers as well?

File: projects/gloo/pkg/plugins/headers/plugin.go [hunk 4]
```diff
@@ -22,7 +24,12 @@ const (
 )
 
 var (
-	MissingHeaderValueError = errors.Errorf("header section of header value option cannot be nil")
+	MissingHeaderValueError = eris.Errorf("header section of header value option cannot be nil")
```
Comment (sam-heilbron) [inline projects/gloo/pkg/plugins/headers/plugin.go:27]: We aren't passing in a value here, should we instead just be using eris.New?

**Review by sam-heilbron — COMMENTED** [review]
Looking good!

File: projects/gloo/pkg/plugins/headers/plugin.go (hunk 2 above)
Comment (sheidkamp) [reply projects/gloo/pkg/plugins/headers/plugin.go:165]: Nope, updating.

File: projects/gloo/pkg/plugins/headers/plugin.go (hunk 4 above)
Comment (sheidkamp) [reply projects/gloo/pkg/plugins/headers/plugin.go]: yup

**Review by sam-heilbron — DISMISSED** [review]

**Comment by sheidkamp** [comment]
> LGTM. I think it makes sense to not attempt to do any type of "best effort" on the header manipulation if any of them are bad. Are the limitations documented?

We point to the envoy documentation in the `headers.proto` file, but it was a dead link. I have updated that link, so users can get to the documentation

**Review by sam-heilbron — APPROVED** [review]

**Review by elcasteel — APPROVED** [review]

**Pull request merged**